File: clients/endpoints/interface.py
import click, sys, json, time, os, math
from SPARQLWrapper import SPARQLWrapper, JSON


def execute(server_url, default_graph_iri, query, timeout):
    sparql_wrapper = SPARQLWrapper(server_url)
    # sparql_wrapper.addDefaultGraph(default_graph_iri)
    sparql_wrapper.setQuery(query)
    sparql_wrapper.setReturnFormat(JSON)
    if timeout is not None:
        sparql_wrapper.setTimeout(timeout)
    try:
        start_time = time.time()
        result = sparql_wrapper.query().convert()
        end_time = time.time()

        execution_time = end_time - start_time
        data_transfer = sys.getsizeof(json.dumps(result))
        nb_solutions = len(result["results"]["bindings"])

        return (result, (execution_time, 1, data_transfer, 0, 0, nb_solutions, 0, 'complete')) 
    except Exception as error:
        end_time = time.time()
        execution_time = end_time - start_time
        if str(error) == 'timed out':
            return ({}, (execution_time, 1, 0, 0, 0, 0, 0, 'timeout'))
        else:
            print(error)
            return ({}, (execution_time, 1, 0, 0, 0, 0, 0, 'error'))
    

@click.group()
def cli():
    pass

@cli.command()
@click.argument("server_url", type=str)
@click.argument("default_graph_iri", type=str)
@click.option('--measure', '-m', type=click.Path(False))
@click.option('--output', '-o', type=click.Path(False))
@click.option('--query', '-q', type=str)
@click.option('--file', '-f', type=click.Path(True, True, False))
@click.option('--timeout', '-t', type=int)
def virtuoso(server_url, default_graph_iri, measure, output, query, file, timeout):
    if query is not None:
        (solutions, (statistics)) = execute(server_url, default_graph_iri, query, timeout)
    elif file is not None:
        query = open(file, 'r').read()
        (solutions, statistics) = execute(server_url, default_graph_iri, query, timeout)
    else:
        print('Error: you must specify a SPARQL query to execute.\nSee interface.py virtuoso --help for more details.')

    if output is not None:
        with open(output, 'w') as output_file:
            output_file.write(json.dumps(solutions))

    (execution_time, nb_calls, data_transfer, data_transfer_approach_overhead, data_transfer_duplicates_overhead, nb_solutions, nb_duplicates, state) = statistics
    
    if measure is not None:    
        with open(measure, 'w') as statistics_file:
            statistics_file.write((
                f'{execution_time},'
                f'{nb_calls},'
                f'{data_transfer},'
                f'{data_transfer_approach_overhead},'
                f'{data_transfer_duplicates_overhead},'
                f'{nb_solutions},'
                f'{nb_duplicates},'
                f'{state}'
            ))

    print((
        'Execution complete !\n'
        f'\t- time: {execution_time} sec\n'
        f'\t- calls: {nb_calls} http requests\n'
        f'\t- transfer: {data_transfer} bytes\n'
        f'\t\t- approach overhead: {data_transfer_approach_overhead} bytes\n'
        f'\t\t- duplicates overhead: {data_transfer_duplicates_overhead} bytes\n'
        f'\t- solutions: {nb_solutions} solution mappings\n'
        f'\t\t- duplicates: {nb_duplicates}\n'
    ))

@cli.command()
@click.argument("server_url", type=str)
@click.argument("default_graph_iri", type=str)
@click.option('--measure', '-m', type=click.Path(False))
@click.option('--output', '-o', type=click.Path(False))
@click.option('--query', '-q', type=str)
@click.option('--file', '-f', type=click.Path(True, True, False))
@click.option('--timeout', '-t', type=int)
def fuseki(server_url, default_graph_iri, measure, output, query, file, timeout):
    if query is not None:
        (solutions, (statistics)) = execute(server_url, default_graph_iri, query, timeout)
    elif file is not None:
        query = open(file, 'r').read()
        (solutions, statistics) = execute(server_url, default_graph_iri, query, timeout)
    else:
        print('Error: you must specify a SPARQL query to execute.\nSee interface.py fuseki --help for more details.')

    if output is not None:
        with open(output, 'w') as output_file:
            output_file.write(json.dumps(solutions))
    
    (execution_time, nb_calls, data_transfer, data_transfer_approach_overhead, data_transfer_duplicates_overhead, nb_solutions, nb_duplicates, state) = statistics

    if measure is not None:    
        with open(measure, 'w') as statistics_file:
            statistics_file.write((
                f'{execution_time},'
                f'{nb_calls},'
                f'{data_transfer},'
                f'{data_transfer_approach_overhead},'
                f'{data_transfer_duplicates_overhead},'
                f'{nb_solutions},'
                f'{nb_duplicates},'
                f'{state}'
            ))

    print((
        'Execution complete !\n'
        f'\t- time: {execution_time} sec\n'
        f'\t- calls: {nb_calls} http requests\n'
        f'\t- transfer: {data_transfer} bytes\n'
        f'\t\t- approach overhead: {data_transfer_approach_overhead} bytes\n'
        f'\t\t- duplicates overhead: {data_transfer_duplicates_overhead} bytes\n'
        f'\t- solutions: {nb_solutions} solution mappings\n'
        f'\t\t- duplicates: {nb_duplicates}\n'
    ))

if __name__ == "__main__":
    cli()

File: clients/sage-multi/bin/interface.js
const program = require('commander')
const fs = require('fs-extra')

const SageClient = require('../src/client').SageClient
const Spy = require('../src/spy').Spy

// ====================================================================================================
// ===== Command line interface =======================================================================
// ====================================================================================================

program.description('Evaluates a set of SPARQL queries using the web preemption model')
    .usage('<server-url> <default-graph-iri> [options]')
    .option('-q, --query <query>', 'Evaluates the given SPARQL query')
    .option('-f, --file <file>', 'Evaluates the SPARQL query in the given file')
    .option('-m, --method <method>', 'Evaluates the SPARQL query using the specified method: [mono, multi]', 'multi')
    .option('-m, --measure <measure>', 'Stores the query execution statistics in the given file')
    .option('-o, --output <output>', 'Stores the query result in the given file')
    .option('-p, --print', 'prints each new result during query execution')
    .option('--timeout <timeout>', 'Sets a time limitation. No more http requests will be sent to the server after the time is out', 0)
    .parse(process.argv)

if (program.args.length !== 2) {
    console.error('Error: missing required arguments ! USAGE: sage-query.js <server-url> <default-graph-iri> [options]')
    process.exit(1)
} else if (!['mono', 'multi', 'alpha'].includes(program.method)) {
    console.error('Error: invalid evaluation strategy. Available strategies are: [mono, multi, direct]')
    console.error('- mono : property paths queries are evaluated using a client-side automaton-based approach with a mono-predicate automaton')
    console.error('- multi : property paths queries are evaluated using a client-side automaton-based approach with a multi-predicate automaton')
} else if (!program.file && !program.query) {
    process.stderr.write('Error: you must specify a SPARQL query to execute.\nSee ./comunica.js --help for more details.\n')
    process.exit(1)
}

// ====================================================================================================
// ===== Query evaluation =============================================================================
// ====================================================================================================

function get_client(server, graph, spy) {
    if (program.method === 'mono') {
        let options = {'property-paths-strategy': 'mono-predicate-automaton'}
        return new SageClient(server, graph, spy, options=options)
    } else {
        let options = {'property-paths-strategy': 'multi-predicate-automaton'}
        return new SageClient(server, graph, spy, options=options)
    } 
}

function eval(server, graph, query) {
    return new Promise((resolve, _) => {
        let spy = new Spy()
        let solutions = []
        let start_time = Date.now()
        let execution_time = 0
        let http_calls = 0
        let data_transfer = 0
        let data_transfer_approach_overhead = 0
        let data_transfer_duplicates_overhead = 0
        let nb_results = 0
        let nb_duplicates = 0
        let state = 'complete'
        get_client(server, graph, spy).execute(query, program.timeout).subscribe(
            (solution) => {
                solutions.push(solution)
                if (program.print) {
                    console.log(solution)
                }
            }, (error) => {
                console.log(error)
                execution_time = (Date.now() - start_time) / 1000.0
                execution_time = (program.timeout > 0 && execution_time > program.timeout) ? program.timeout : execution_time
                http_calls = spy.nb_http_calls
                data_transfer = spy.data_transfer
                data_transfer_approach_overhead = Math.abs(data_transfer - Buffer.byteLength(JSON.stringify(solutions), 'utf8'))
                nb_results = solutions.length
                let statistics = [
                    execution_time, 
                    spy.nb_http_calls, 
                    spy.data_transfer, 
                    data_transfer_approach_overhead, 
                    data_transfer_duplicates_overhead, 
                    nb_results, 
                    nb_duplicates, 
                    'error']
                resolve([solutions, statistics])
            }, () => {
                execution_time = (Date.now() - start_time) / 1000.0
                execution_time = (program.timeout > 0 && execution_time >= program.timeout) ? program.timeout : execution_time
                http_calls = spy.nb_http_calls
                data_transfer = spy.data_transfer
                data_transfer_approach_overhead = Math.abs(data_transfer - Buffer.byteLength(JSON.stringify(solutions), 'utf8'))
                nb_results = solutions.length
                state = (program.timeout > 0 && execution_time >= program.timeout) ? 'timeout' : 'complete'
                let statistics = [
                    execution_time, 
                    spy.nb_http_calls, 
                    spy.data_transfer, 
                    data_transfer_approach_overhead, 
                    data_transfer_duplicates_overhead, 
                    nb_results, 
                    nb_duplicates, 
                    state]
                resolve([solutions, statistics])
            }
        )
    })
}

async function execute(server, graph, query) {
    let [solutions, statistics] = await eval(server, graph, query)
    // Write the query execution statistics in the given file
    if (program.measure) {
        let data = statistics.join(',') + '\n'
        fs.writeFileSync(program.measure, data, {encoding: 'utf-8'})
    }
    // Writes the query result in the given file
    if (program.output) {
        let data = JSON.stringify(solutions)
        fs.writeFileSync(program.output, data, {encoding: 'utf-8'})
    }
    // Prints query execution statistics
    console.log(`Execution complete !
    - time: ${statistics[0]} sec
    - calls: ${statistics[1]} http requests
    - transfer: ${statistics[2]} bytes
        - approach overhead: ${statistics[3]} bytes
        - duplicates overhead: ${statistics[4]} bytes
    - solutions: ${statistics[5]} solution mappings
        - duplicates: ${statistics[6]}`)
}

if (program.file) {
    execute(program.args[0], program.args[1], fs.readFileSync(program.file).toString()) 
} else {
    execute(program.args[0], program.args[1], program.query)
} 